public/app/features/correlations/Forms/QueryEditorField.tsx

```diff
@@ -25,6 +49,29 @@ export const QueryEditorField = ({ dsUid, invalid, error, name }: Props) => {
   }, [dsUid]);
   const QueryEditor = datasource?.components?.QueryEditor;
 
+  const handleValidation = async (value: DataQuery) => {
+    if (datasource) {
+      runner.run({
+        datasource: datasource,
+        queries: [value],
+        timezone: 'utc',
+        timeRange: getDefaultTimeRange(),
+        maxDataPoints: 100,
+        minInterval: null,
+      });
+
+      await runner.get().subscribe((panelData) => {
+        if (!panelData || panelData.state === 'Error') {
+          setIsValidQuery(false);
+        } else if (panelData.state === 'Done') {
+          setIsValidQuery(true);
+        } else {
+          setIsValidQuery(undefined);
+        }
+      });
+    }
+  };
```

Comment:
we don't need `handleValidation` to be marked as async (`subscribe` does not return a promise/we don't need to use its return value) 
```suggestion
  const handleValidation = (value: DataQuery) => {
    if (datasource) {
      runner.run({
        datasource: datasource,
        queries: [value],
        timezone: 'utc',
        timeRange: getDefaultTimeRange(),
        maxDataPoints: 100,
        minInterval: null,
      });

      runner.get().subscribe((panelData) => {
        if (!panelData || panelData.state === 'Error') {
          setIsValidQuery(false);
        } else if (panelData.state === 'Done') {
          setIsValidQuery(true);
        } else {
          setIsValidQuery(undefined);
        }
      });
    }
  };
```

Reply:
also (this is more of a nit than anything) we could wrap it in a useCallback so it doesn't get recreated on every render (
```suggestion
  const handleValidation = async (value: DataQuery) => {
    if (datasource) {
      runner.run({
        datasource: datasource,
        queries: [value],
        timezone: 'utc',
        timeRange: getDefaultTimeRange(),
        maxDataPoints: 100,
        minInterval: null,
      });

      await runner.get().subscribe((panelData) => {
        if (!panelData || panelData.state === 'Error') {
          setIsValidQuery(false);
        } else if (panelData.state === 'Done') {
          setIsValidQuery(true);
        } else {
          setIsValidQuery(undefined);
        }
      });
    }
  };
```

Reply:
1. I agree and I'll change it.
2. I don't understand yet why this should get recreated with rerendering. `handleValidation()` is only called when the related button gets clicked and the QueryRunner itself is only created once as it is already wrapped in `useMemo()`

Reply:
this is not about calling `handleValidation` but defining the function itself. every time the state or a prop changes the component rerenders - and by doing so recreates the `handleValidation` function. this happens every time you click on `Validate Query`.

but anyway it's not a biggie, we pass an anonymous function to the validate button anyway so it won't make much of a difference.

Reply:
there's another issue with this though, by subscribing every time we run the validation (by calling `runner.get().subscribe(..)` we are basically creating a subscription every time you click on the button, which is a memory leak. if you put a console statement inside the subscriber you'll see how it runs once more every time you click.

we should probably move the subscription somewhere else and make sure to unsubscribe when needed. i think something like
```ts
  useEffect(() => {
    runner.get().subscribe((panelData) => {
      console.log(panelData);
      if (!panelData || panelData.state === 'Error') {
        setIsValidQuery(false);
      } else if (panelData.state === 'Done') {
        setIsValidQuery(true);
      } else {
        setIsValidQuery(undefined);
      }
    });

    return runner.destroy;
  }, [runner]);
```
(outside of the `handelValidation` function would suffice)

Reply:
I see why this needs to be changed. I didn't expect so many subscriptions and just changed it to what you suggested.